GC latency improvements

damiendoligez opened the pull request:
This patch introduces several changes to the GC, mostly aimed at reducing worst-case latency:
1. When a GC cycle starts, the global roots must be marked. This is now done incrementally. (asmrun/roots.c -- native-code only because byte-code doesn't have this problem, it has only one global root).
2. The minor GC and major GC slice are separated and the major slice is run at the mid-point between minor collections (i.e. when the minor heap is half-full).
3. Major GC slice size is now (under user control) amortized over several minor collections instead of varying directly with the promotion rate.
4. Large arrays are now marked incrementally instead of blocking the program for a long time.
5. Allocation of major heap chunks is (optionally) done with `mmap` using the HUGE_PAGES feature to lower the TLB load on large heap sizes.

minor improvements:
- better control of GC verbosity in debug runtime
- added a runtime variant with GC performance instrumentation
- missing include in instrtrace.c
- add an inline function for adding an entry to a `ref_table`
- add functions to stdlib modules `Gc` and `Sys` to control the new features
- better display of GC stats

kit-ty-kate commented:
Have you tried to squash all the commits, stash the resulting commit, pull from trunk and apply the stashed changes ? Maybe it can work.

gasche commented:
We discussed this large change at the developer meeting yesterday, and decided to include it for 4.03 -- the latency improvement are rather impressive. Below is a quick summary of the salient points of the discussion.

The flambda people in the room ( @chambart and @mshinwell ) seemed nervous at the mention that the work on incremental marking of GC roots relied on their immutability, so they may want to check the details with Damien.

One thing that was forgotten in the above description is the incremental marking of large arrays.

One other thing that is unclear in the above description is the status of the `mmap` tricks. My understanding is that they are _disabled_ by default, and can be enabled (but is it a configure-time flag or a program-runtime flag?).

There was discussion of the interaction with @mshinwell 's more ambitious no-naked-pointers work in #203 . One idea mentioned by Damien was to use the `mmap` trick in addition to `no-naked-pointers`, to serve as a "debug mode" making it easy to check that the naked accesses are indeed in the heap (but maybe just keeping the pagetable around for this debug mode would suffice?).

damiendoligez committed:
GC latency improvements

damiendoligez commented:
I have updated the patch to turn it into a single commit, and to remove the mmap code. (BTW the mmap option was neither configure-time nor run-time, but compile-time). The consensus was that `no-naked-pointers` is the right thing to do, even if the migration path is more dangerous.

Indeed, the mmap trick is not needed to get a debug mode for `no-naked-pointers`.

I have updated the description above to reflect the current state of the patch (and add the missing bit about large arrays).

damiendoligez commented:
@bobot (and everyone else who would volunteer)
This is now ready for review.

mshinwell commented:
@gasche Nitpick on no-naked-pointers: despite the name (which should maybe be fixed), pointers outside the heap are permitted.  It's just that they need to point at something that looks like an OCaml value.

damiendoligez commented:
The complete `no-naked-pointers` rule is that every pointer to outside the heap must be at least one of the following:
- odd (i.e. look like an int)
- stored in a closure
- to something that has a black header (i.e. the word before the pointed thing must be readable and have the right values in its two color bits)

mshinwell commented:
As I have mentioned to @damiendoligez in private email, I suspect there is a bug in this patch that causes it to fail when merged with the flambda branch.  It's probably to do with the multiple roots thing (GPR #262).

xavierleroy commented on asmcomp/amd64/emit.mlp:
```diff
@@ -276,6 +277,7 @@ let call_gc_sites = ref ([] : gc_call list)
 
 let emit_call_gc gc =
   def_label gc.gc_lbl;
+  I.sub (int gc.gc_size) r15;
```
Why "sub" ?  I was naively expecting "add".

damiendoligez replied:
Indeed, and this begs the question: why didn't it get caught in testing? With the current design, it's very unlikely to cause problems because the pointer is (almost) never read. I need to brainstorm with you about this.

damiendoligez committed:
fix huge pages vs renaming of caml_stat_heap_size

damiendoligez commented on byterun/caml/address_class.h:
```diff
@@ -16,12 +16,18 @@
+#define Is_in_heap(a) (Classify_addr(a) & In_heap)
+#define Is_in_heap_or_young(a) (Classify_addr(a) & (In_heap | In_young))
+#define Is_in_value_area(a)                                     \
+  (Classify_addr(a) & (In_heap | In_young | In_static_data))
```
merging mistake

damiendoligez commented on byterun/caml/minor_gc.h:
```diff
@@ -34,9 +36,18 @@ struct caml_ref_table {
+    *(tbl).ptr++ = (p);                           \
+  } while(0)
+
 extern void caml_set_minor_heap_size (asize_t); /* size in bytes */
```
Will do.

damiendoligez commented on stdlib/gc.ml:
```diff
@@ -38,6 +38,7 @@ type control = {
   mutable max_overhead : int;
   mutable stack_limit : int;
   mutable allocation_policy : int;
+  window_size : int;
```
The others are mutable because when this record was introduced, the record update  feature `{ r with lbl = e }` didn't exist. You would `get`, mutate the result, and use it to `set`. Now you simply `set {(get()) with window_size = 5}`.

If I could add a "deprecated" annotation on these `mutable` flags, I would. Since it is deprecated, I see no reason to duplicate it in new fields. I can change it if you think it's too weird.

alainfrisch replied:
> If I could add a "deprecated" annotation on these mutable flags

This is tempting!  That would probably be an annotation on the field (e.g. `[@deprecated_mutable]`).

alainfrisch replied:
Done!  One can now write:

``` ocaml
type control = {
   mutable minor_heap_size : int
       [@@mutable_deprecated "Use {ctrl with minor_heap_size = ...} instead."];
   ...
```

I didn't actually do the change in Gc for now.

damiendoligez commented on byterun/gc_ctrl.c:
```diff
@@ -511,6 +590,54 @@ void caml_init_gc (uintnat minor_size, uintnat major_size,
+    ("a=%d,b=%s,H=%lu,i=%lu,l=%lu,o=%lu,O=%lu,p=%d,s=%lu,t=%d,v=%lu,w=%d,W=%lu",
+     /* a */ caml_allocation_policy,
+     /* b */ caml_backtrace_active,
+     /* h */ /* missing */ /* FIXME add when changed to min_heap_size */
```
Yes. It's another GC feature that I'll introduce later.

damiendoligez committed:
fix sign error in GC call

damiendoligez committed:
remove duplicate macro definitions

damiendoligez committed:
deduplicate definition of `inline`

damiendoligez committed:
transform `Add_to_ref_table` macro into inline function

damiendoligez committed:
add comments

damiendoligez committed:
fix bug in incremental marking of roots

damiendoligez commented:
I've made changes according to the remarks of the reviewers, and fixed a nasty bug in the root scanning (triggered by  testing with flambda).

Following a discussion with @xavierleroy I'm going to remove the change to the protocol for calling the GC.

damiendoligez committed:
Revert the change to the GC calling protocol: the allocation pointer
does not need to point to a valid header when the GC is called. On the
other hand, the major GC can only start a cycle when the minor heap
(allocation arena) is empty.

damiendoligez committed:
ignore generated files for the instrumented runtime

damiendoligez commented:
@bobot could you please review this version with particular attention to `asmrun/roots.c`? It's quite tricky.

damiendoligez commented on asmrun/roots.c:
```diff
@@ -335,27 +335,89 @@ void caml_oldify_local_roots (void)
+      incr_roots_sz = Wosize_val (*incr_roots_glob);
+      for (incr_roots_j = 0; incr_roots_j < incr_roots_sz; incr_roots_j++){
+        goto setup_done;
+      }
```
The idea here is to have an exact copy of the loops found in `caml_do_roots`. Maybe I should add a comment to this effect.

damiendoligez commented:
> Do you have a test for the fix in def7f4d ?

It's a complete install of the compiler with flambda merged in... not really usable as a regression test.

damiendoligez committed:
fix wrong variable name caml_stat_heap_size

damiendoligez committed:
simplify incremental root scanning

damiendoligez commented:
Updated with @bobot's suggestions and fixes. If anyone else wants to review, now's the time.

damiendoligez committed:
bootstrap for new primitive

mrvn commented:
1) If no naked pointer are allowed then where is the helper function to allocate a chunk of memory outside the ocaml heap with a suitable black colored header? Something like

  CAMLextern void \* caml_alloc_unmovable(mlsize_t); /\* allocate a block of memory outside the ocaml heap suitable for naked C/C++ objects _/
  CAMLextern void caml_free_unmovale(void *); /_ free block allocated by caml_alloc_unmovable */

2) I frequently have C objects that also need to hold some values. At the moment one has to allocate an ocaml block to hold the values and malloc for the C object. Without naked pointers now one further needs a second ocaml block with No_scan_tag (or the alloc from 1).  Instead couldn't we have a single block outside the ocaml heap containing the values at the start and the C structure at the end where the Gc scans only the begining?

  /\* allocate a block of memory outside the ocaml heap with mixed content
- Allocate a block with tag <tag> containing <values> ocaml values that
- will be scanned by the Gc if reachable, followed by <no_scan> bytes of
- of memory that will be ignored by the Gc. The block will never be moved
- by the Gc and the no_scan portion can be accessed without holding the
- runtime lock. The values portion must only be accessed while holding the
- runtime lock.
  */
  CAMLextern void \* caml_alloc_unmovable(mlsize_t values, tag_t tag, mlsize_t no_scan);

3) Alternatively maybe the 'struct custom_operations' could be extended to include a field saying how many words to scan in the custom block. That way one could allocate single blocks containing both ocaml values and raw data and still have them movable by the Gc.

damiendoligez commented:
I think (1) is a good idea. cc @mshinwell 

For (2): if it's outside the heap and points into the heap, then it has to be a root. It would probably not reasonable to have too many such blocks. We don't currently have a data structure to register blocks of roots, so this option would need big changes to globroots.c.

(3) looks like a good idea, but we have to be careful to remain compatible with C code that uses the current version.

let-def commented:
Hijacking a bit the conversation, but related to C-FFI and GC:

When dealing with C-FFI needing a back-reference to OCaml values, I felt that a special area in the OCaml heap that is guaranteed to be non-compacted (and as such never moved) would help a lot.
This area would behave exactly like the major heap (for collection), but without compaction. Values would be allocated via a specific function.

The recent [PR 7162](http://caml.inria.fr/mantis/view.php?id=7162) reminded me of this. Exchanging the idea with @garrigue, he thought this could globally benefit LablGTK (not having to disable compaction for the whole program).
@yminsky also suggested this could enable working on some OCaml values from C-FFI without locking the runtime (e.g. having C-like arrays / bigstrings in the heap).

@damiendoligez do you think this is reasonable?

damiendoligez commented:
@bobot I've thought of that but won't it be too slow?

@def-lkb It looks doable although it'll need a fair amount of work. Could you open a feature request on Mantis?

jhjourdan commented:
@damiendoligez Instead of a general function, we could have a bitset describing which words are pointers and which are raw data.

mrvn commented:
@damiendoligez For 2) allocate a chunk of memory, put a block with N values at the start, add one value pointing at itself and register it as root, remaining space is for C use. The free function would unregister the root and free the block. It's what everyone already does now except that the allocations get merged into one block and the registering is automatic.

@def-lkb I'm not sure a special unmovable heap would be that beneficial. You could only use it from C and with values a module allocates itself because the type system doesn't know about unmovables. And you still have to register a root to keep the value alive. You might as well just allocate the object fully outside the ocaml heap (there should be helper functions for this creating the block header). With the no-naked-pointer change I think that would do exactly what you want without ocaml having a special unmovable heap. Secondly accessing such a block without the runtime lock would only work if all values in the block only point to other unmovable objects, which basically rules out mutable fields. You could only do that in a verry controlled structure that you set up with a deep copy of all fields.

mrvn commented:
@damiendoligez 

> It would probably not reasonable to have too many such blocks.

Is 10-1000 appearing and disapearing per second too many? In my QT5 bindings if you do catch for example mouse events and then react to them by calling Qt functions that return anything but primitive types you get a ton of allocationd and eventual garbage collections verry quickyly. At the moment I register them as generational global root and the referenced objects generally never make it out of the minor heap. Hopefully that means the roots get unregistered before a major collection and their numbers doesn't matter. Right?

DemiMarie commented:
One feature I would like to see is "mixed blocks": blocks that are in the ordinary OCaml heap, but have only part of themselves scanned by the GC.  Applications include:
- Unboxing of int32, int64, nativeint, double when in records/tuples that also contain other OCaml values
- Resizable arrays (only the part that is in use needs to be scanned)
- Storing C-allocated data (such as a callback pointer) in an OCaml block without requiring multiple allocations

@mrvn Accessing _unboxed_ data will still work if you know (through external invariants) that no OCaml thread is accessing it.  Applications include zero-copy writing of data to disk or the network.

damiendoligez commented:
@mrvn 
It's not a question of how many appear and disappear, but how many are present at any time.
Anyway, your use case looks reasonable.

alainfrisch commented:
I suspect this PR is the cause for http://caml.inria.fr/mantis/view.php?id=7228

damiendoligez commented on stdlib/gc.ml:
```diff
@@ -38,6 +38,7 @@ type control = {
   mutable max_overhead : int;
   mutable stack_limit : int;
   mutable allocation_policy : int;
+  window_size : int;
```
FTR, this is commit 6b9d72d532c69aed263497b1cadcc2742cb20a73 and the syntax is `[@deprecated_mutable "..."]` or `[@ocaml.deprecated_mutable "..."]`.
